crates/net/downloaders/src/headers/reverse_headers.rs

```diff
@@ -211,59 +252,45 @@ where
-                                    received: parent.number,
-                                    expected: number,
-                                },
-                            })
-                        }
-                    }
-                    SyncTargetBlock::HashAndNumber { hash, .. } => {
-                        if parent.hash() != hash {
-                            return Err(HeadersResponseError {
-                                request,
-                                peer_id: Some(peer_id),
-                                error: DownloadError::InvalidTip {
-                                    received: parent.hash(),
-                                    expected: hash,
-                                },
-                            })
-                        }
-                    }
-                }
+                self.validate_sync_target(&parent, request.clone(), peer_id)?;
             }
 
             validated.push(parent);
         }
 
         // If the last (smallest) validated header attaches to the local head, validate it.
-        if let Some((last_header, head)) = validated.last().zip(self.local_head.as_ref()) {
-            if last_header.number == head.number + 1 {
-                if let Err(error) = self.validate(last_header, head) {
-                    trace!(target: "downloaders::headers", ?error ,"Failed to validate header");
-                    return Err(HeadersResponseError { request, peer_id: Some(peer_id), error })
+        if let Some((last_header, head)) = validated
+            .last_mut()
+            .zip(self.local_head.as_ref())
+            .filter(|(last, head)| last.number == head.number + 1)
+        {
+            // Every header must be valid on its own
+            if let Err(error) = self.consensus.validate_header(last_header) {
+                trace!(target: "downloaders::headers", ?error, "Failed to validate header");
+                return Err(HeadersResponseError {
+                    request,
+                    peer_id: Some(peer_id),
+                    error: DownloadError::HeaderValidation { hash: head.hash(), error },
+                }
+                .into())
+            }
+
+            // If the header is valid on its own, but not against it's parent, we return it an error
+            // `HeaderResponse::Detached`.
+            if let Err(error) = self.consensus.validate_header_against_parent(last_header, head) {
+                // Replace the last header with a detached variant
+                error!(target: "downloaders::headers", ?error, number = last_header.number, hash = ?last_header.hash, "Header cannot be attached to known canonical chain");
+                return Err(HeadersDownloaderError::HeaderCannotBeAttached {
+                    number: last_header.number,
+                    hash: last_header.hash,
+                    error,
                 }
+                .into())
             }
         }
```

Comment:
this is the crucial check that ensures that the local head is a parent of the downloaded (validated) header.